File: parser.ml
open Printf

type animal = Dog of string | Cat of string

let say x =
    match x with
    | Dog x -> x ^ " says woof"
    | Cat x -> x ^ " says meow"
;;


let said = say (Cat "Fluffy ") ;;

let () = printf "%s" said;;
let () = print_endline ""
let () = print_endline ""
let () = printf "%s" (say (Dog "Jimmy "))
let () = print_endline ""
let () = print_endline ""
let jimmy = Dog "Jimmy"
let () = printf "%s" (say jimmy)
let () = print_endline ""
let () = print_endline ""

let () = printf "%s" ((func -> (jimmy x) -> x ) ^ " test")


(*  
let in_file = "header.hcaml"
let out_file = "header.html"
let message = "Hello!"
  
let () =
  (* Write message to file *)
  let oc = open_out out_file in    (* create or truncate file, return channel *)
  fprintf oc "%s\n" message;   (* write something *)   
  close_out oc;                (* flush and close the channel *)
  
  (* Read file and display the first line *)
  let ic = open_in in_file in
  try 
    let line = input_line ic in  (* read line from in_channel and discard \n *)
    print_endline line;          (* write the result to stdout *)
    flush stdout;                (* write on the underlying device now *)
    close_in ic                  (* close the input channel *) 
  
  with e ->                      (* some unexpected exception occurs *)
    close_in_noerr ic;           (* emergency closing *)
    raise e                      (* exit with error: files are closed but
                                    channels are not flushed *)
  
  (* normal exit: all channels are flushed and closed *)
  *)

File: parser.py
import re
in_file = "header.hcaml"
out_file = "header.html"

f = open(in_file, "r")

# print(f.read())

source_code = f.read()


def parser(input):

    # Current position

    _tokens = []
    _current = 0
    while _current < len(input):
        _char = input[_current]

        if (_char == "[" or _char == "]"):
            _tokens.append({ "type": "paren", "value": _char})
            _current += 1
            continue



        SPACE = " "
        if (re.search(SPACE, _char)):
            _value = ""
            while (re.search(SPACE, input[_current])):
                _value = _value + _char
                _current  = _current + 1 

            continue

        LETTERS = "[a-z]|[A-Z]|[0-9]"
        if (re.search(LETTERS, _char)):
            _value = ""
            count = _current
            while (re.search(LETTERS, input[_current])):
                _value = _value + input[_current]
                _current = _current + 1 
                count 

            _tokens.append({"type": "name", "value": _value})
            continue

        
        _current += 1
    return _tokens

token = parser(source_code)

print(token)
        



